core/graphelier-service/db/mongo.go

```diff
@@ -29,20 +28,64 @@ func NewConnection() (*Connector, error) {
 	if err = client.Ping(context.TODO(), nil); err != nil {
 		log.Fatal(err)
 	}
-	fmt.Println("Connected to MongoDB :)")
+	log.Println("Connected to MongoDB :)")
 
 	return &Connector{client}, nil
 }
 
-// FindOrderbook : Finds the Orderbook of an instrument based on the timestamp requested in the db
-func (c *Connector) FindOrderbook(instrument string, timestamp float64) (result *models.Orderbook) {
+// GetOrderbook : Finds the Orderbook of an instrument based on the timestamp requested in the db
+func (c *Connector) GetOrderbook(instrument string, timestamp uint64) (result *models.Orderbook) {
 	collection := c.Database("graphelier-db").Collection("orderbooks")
-	filter := bson.D{{Key: "instrument", Value: instrument}, {Key: "timestamp", Value: timestamp}}
+	filter := bson.D{
+		{Key: "instrument", Value: instrument},
+		{Key: "timestamp", Value: bson.D{
+			{Key: "$lte", Value: timestamp},
+		}},
+	}
+	option := options.FindOne()
+	option.SetSort(bson.D{{Key: "timestamp", Value: -1}})
 
-	err := collection.FindOne(context.TODO(), filter).Decode(&result)
+	err := collection.FindOne(context.TODO(), filter, option).Decode(&result)
 	if err != nil {
 		log.Fatal(err)
 	}
 
 	return result
 }
+
+// GetMessages : Finds the Message of an instrument based on the timestamp requested
+func (c *Connector) GetMessages(instrument string, timestamp uint64, latestFullSnapshot uint64) (results []*models.Message) {
+	collection := c.Database("graphelier-db").Collection("messages")
+	filter := bson.D{
+		{Key: "instrument", Value: instrument},
+		{Key: "timestamp", Value: bson.D{
+			{Key: "$lte", Value: timestamp},
+			{Key: "$gte", Value: latestFullSnapshot},
+		}},
+	}
+
+	findOptions := options.Find()
+
+	cursor, err := collection.Find(context.TODO(), filter, findOptions)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for cursor.Next(context.TODO()) {
+		var m models.Message
+		err := cursor.Decode(&m)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		results = append(results, &m)
+	}
+
+	if err := cursor.Err(); err != nil {
+		log.Fatal(err)
+	}
+
+	cursor.Close(context.TODO())
```

Comment:
In GO you can use the keyword `defer` to have a line execute after a function's return.
People use it to close cursors and connections all the time. So instead of having this at line 88, you can write the following right after line 69.
```golang
defer cursor.Close(context.TODO())
```
You don't need to change this by the way. What you did and using defer are equivalent. I'm just putting this info out there for future use.